lib/IntermediateRepresentation/mixin.h

```diff
@@ -0,0 +1,114 @@
+  SumTypeMixin(const SumTypeMixin &) = delete;
+  SumTypeMixin &operator=(const SumTypeMixin &) = delete;
+  SumTypeMixin() = delete;
+  T u;
+};
+
+template<typename T, typename E = void> struct SumTypeCopyMixin {};
+template<typename T>  // T must be std::optional<...>
+struct SumTypeCopyMixin<T, std::enable_if_t<std::variant_size_v<T>>> {
+  using CopyableSumTypeTrait = std::true_type;
+  SumTypeCopyMixin(SumTypeCopyMixin &&) = default;
+  SumTypeCopyMixin &operator=(SumTypeCopyMixin &&) = default;
+  SumTypeCopyMixin(const SumTypeCopyMixin &) = default;
+  SumTypeCopyMixin &operator=(const SumTypeCopyMixin &) = default;
+  SumTypeCopyMixin() = delete;
+  T u;
+};
+#define SUM_TYPE_COPY_MIXIN(Derived) \
+  Derived(const Derived &derived) : SumTypeCopyMixin(derived) {} \
+  Derived &operator=(const Derived &derived) { \
+    SumTypeCopyMixin::operator=(derived); \
+    return *this; \
+  }
+
+template<typename T, typename E = void> struct ProductTypeMixin {};
+template<typename T>  // T must be std::tuple<...>
+struct ProductTypeMixin<T, std::enable_if_t<std::tuple_size_v<T>>> {
+  template<typename A> ProductTypeMixin(A &&x) : t{std::move(x)} {}
+  using ProductTypeTrait = std::true_type;
+  ProductTypeMixin(ProductTypeMixin &&) = default;
+  ProductTypeMixin &operator=(ProductTypeMixin &&) = default;
+  ProductTypeMixin(const ProductTypeMixin &) = delete;
+  ProductTypeMixin &operator=(const ProductTypeMixin &) = delete;
+  ProductTypeMixin() = delete;
+  T t;
+};
+
+template<typename T, typename E = void> struct MaybeMixin {};
+template<typename T>  // T must be std::optional<...>
+struct MaybeMixin<T,
+    std::enable_if_t<
+        std::is_same_v<std::optional<typename T::value_type>, T>>> {
+  template<typename A> MaybeMixin(A &&x) : o{std::move(x)} {}
+  using MaybeTrait = std::true_type;
+  MaybeMixin(MaybeMixin &&) = default;
+  MaybeMixin &operator=(MaybeMixin &&) = default;
+  MaybeMixin(const MaybeMixin &) = delete;
+  MaybeMixin &operator=(const MaybeMixin &) = delete;
+  MaybeMixin() = delete;
+  T o;
+};
+
+template<typename T, typename P> struct ChildMixin {
+protected:
+  P *parent;
+
+public:
+  ChildMixin(P *p) : parent{p} {}
+  inline const P *getParent() const { return parent; }
+  inline P *getParent() { return parent; }
+  llvm::iplist<T> &getList() { return parent->getSublist(this); }
+};
+
+template<typename A, typename B, typename C>
+C Zip(C out, A first, A last, B other) {
+  std::transform(first, last, other, out,
+      [](auto &&a, auto &&b) -> std::pair<decltype(a), decltype(b)> {
+        return {a, b};
+      });
+  return out;
+}
+template<typename A, typename B> B &Unzip(B &out, A first, A last) {
+  std::transform(first, last, std::back_inserter(out.first),
+      [](auto &&a) -> decltype(a.first) { return a.first; });
+  std::transform(first, last, std::back_inserter(out.second),
+      [](auto &&a) -> decltype(a.second) { return a.second; });
+  return out;
+}
+
+}
+
+#endif
```

Comment:
Please put a comment on the terminal `#endif` in header files that matches the guard at the top of the file.

Reply:
Done